10_cdev/chardev.c

```diff
@@ -0,0 +1,366 @@
+		class_destroy(sys_class);
+		return -EINVAL;
+	}
+
+	res = class_create_file(sys_class, &class_attr_clear);
+	if (res != 0) {
+		pr_err("bad file 'total_characters_processed' create\n");
+		class_destroy(sys_class);
+		return -EINVAL;
+	}
+
+	pr_info("sys_char class in sysfs initialized\n");
+	return res;
+}
+
+void clean_sys_file(void)
+{
+	class_remove_file(sys_class, &class_attr_clear);
+	class_remove_file(sys_class, &class_attr_numop);
+	class_destroy(sys_class);
+}
+
+static int __init char_init(void)
+{
+	int err;
+	if (length > LENGTH_DEF) {
+		pr_err("Wrong length of buffer");
+		return -EINVAL;
+	}
+
+	pr_info("Initializing task10 module\n");
+
+	majorNumber = register_chrdev(0, DEVICE_NAME, &fops);
+	if (majorNumber < 0) {
+		pr_err("Failed to register a major number\n");
+		return majorNumber;
+	}
+
+	pr_info("Registered correctly with major number %d\n", majorNumber);
+
+	charClass = class_create(THIS_MODULE, CLASS_NAME);
+	if (IS_ERR(charClass)) {
+		unregister_chrdev(majorNumber, DEVICE_NAME);
+		pr_err("Failed to register device class\n");
+		return PTR_ERR(charClass);
+	}
+
+	pr_info("Device class registered correctly\n");
+
+	charDevice = device_create(charClass, NULL,
+				MKDEV(majorNumber, 0), NULL, DEVICE_NAME);
+
+	if (IS_ERR(charDevice)) {
+		class_destroy(charClass);
+		unregister_chrdev(majorNumber, DEVICE_NAME);
+		pr_err("Failed to create the device\n");
+		return PTR_ERR(charDevice);
+	}
+
+	message = kmalloc(length * sizeof(char), GFP_KERNEL);
+	if (message == NULL)
+		return -ENOMEM;
+
+	pr_info("Buffer size is %d\n", length);
+	pr_info(" Char device class created correctly\n");
+
+	err = create_buffer();
+	if (err)
+		goto error;
+
+	err = create_proc_dir();
+	if (err)
+		goto error;
+	pr_notice("Procfs volume and size loaded\n");
+
+	err = create_sys_file();
+	if (err)
+		goto error;
+
+	return 0;
+
+error:
+	pr_err("Failed to load\n");
+	cleanup_proc();
+	cleanup_buffer(volume_buffer);
+	cleanup_buffer(size_buffer);
+	return err;
+}
+
+static void __exit char_exit(void)
+{
+	device_destroy(charClass, MKDEV(majorNumber, 0));
+	class_unregister(charClass);
+	class_destroy(charClass);
+	unregister_chrdev(majorNumber, DEVICE_NAME);
+	kfree(message);
+	cleanup_proc();
+	cleanup_buffer(volume_buffer);
+	cleanup_buffer(size_buffer);
+	clean_sys_file();
+	pr_info("Bye!\n");
+}
+
+static int dev_open(struct inode *inodep, struct file *filep)
+{
+	numberOpens++;
+	pr_info("Device has been opened %d times\n", numberOpens);
+	return 0;
+}
+
+static ssize_t dev_read(struct file *filep, char *buffer, size_t len,
+			loff_t *offset)
+{
+	int ret = 0;
+
+	if (*offset)
+		return 0;
+
+	ret = copy_to_user(buffer, message, size_of_message);
+	if (ret != 0)
+		return -EFAULT;
+
+
+	*offset = size_of_message;
+
+	return *offset;
+}
+
+static ssize_t dev_write(struct file *filep, const char *buffer, size_t len,
+			loff_t *offset)
+{
+	int ret;
+
+	if (len > sizeof(message) - 1)
+		size_of_message = sizeof(message) - 1;
```

Comment:
According to your code, message is a pointer.
What would be the value sizeof(message) ?